Jenkins 4.0.0 Release.

kaiwalyajoshi opened the pull request:
There are many significant updates in this release.
- Use of [USI](https://github.com/mesosphere/usi) as the scheduler for Mesos-Plugin.
- Support for Windows agents.
- Removal of `config.xml` in favour of JCasC.
- Ability to specify a user defined list of plugins to install.
- Base-Tech update to Jenkins 2.204.2 LTS
- Many cleanups for easier maintainabilty and stability.

kaiwalyajoshi committed:
Make MESOS the default containerizer.

kaiwalyajoshi committed:
Make nobody the default user.

kaiwalyajoshi committed:
Purge all unncessary plugins except for a minimal set.

kaiwalyajoshi committed:
Configureable Jenkins Plugins Installation. (#312)

* User configureable list of Jenkins plugins to be installed at service deployment time.

* Reflect README with the new list of bundled plugins.

kaiwalyajoshi committed:
Bump JCasC to 1.35

kaiwalyajoshi committed:
Replace config.xml with JCasC variant. Set Mesos to use mesos.leader instead of retreiving it from Zk.

kaiwalyajoshi committed:
Add missing JCasC file from previous commit.

kaiwalyajoshi committed:
Remove pinned-storage related options.

kaiwalyajoshi committed:
Remove pinned storage-hostnames.

kaiwalyajoshi committed:
Remove Marathon-LB related configs.

kaiwalyajoshi committed:
Re-add volumes support. Split configuration into DC/OS Service, Jenkins-Master and Jenkins-Agent sections respectively.

kaiwalyajoshi committed:
Add support for windows and linux jenkins agents. Remove support for additonal plugins (its not stable enough yet)

jeschkies committed:
Load Mesos plugin from Jenkins CI repo. (#313)

Summary:
The Mesos plugin should be installed from the stable or experimental
repository.

kaiwalyajoshi committed:
Move quota enforcement from scripts/bootstrap.py to an external shell-script as a python process can't alter the environment of its calling shell.

kaiwalyajoshi committed:
Apply correct DNS name when using frameworks in nested marathon groups.

kaiwalyajoshi committed:
Remove Nginx configuration setup from bootstrap.py.

kaiwalyajoshi committed:
Export env-var, not the value of the env-var.

kaiwalyajoshi committed:
Move creation of ssh-known-scripts out of scripts/bootstrap.sh

kaiwalyajoshi committed:
Remove scripts/boostrap.py and with it the Python dependencies. Set default user to nobody.

kaiwalyajoshi committed:
Fix missing entries since options restructuring.

kaiwalyajoshi committed:
Update base-tech to Jenkins 2.204.2 as the current release has security-vulnerabilities.

kaiwalyajoshi committed:
Re-add Marathon-LB settings.

kaiwalyajoshi committed:
Readd configureable Jenkins Plugins Installation.

kaiwalyajoshi committed:
Mention DCOS_OSS-5906 in the Dockerfile for future cleanups & simplifications.

kaiwalyajoshi committed:
Remove unused config/jenkins/config.xml and scripts/bootstrap.py files.

jeschkies commented on conf/jenkins/configuration.yaml:
```diff
@@ -0,0 +1,55 @@
+            agentCommandStyle: Linux
+            containerInfo:
+              dockerForcePullImage: false
+              dockerImage: "${JENKINS_LINUX_AGENT_IMAGE:-mesosphere/jenkins-dind:0.6.0-alpine-experimental}"
```
Out of curiosity, why  are we using a dind image as default and not a simple official JDK image?

kaiwalyajoshi replied:
The common use case for customers is to build docker images on the jenkins agents, the simple JDK image doesn't do this and will break the default workflow for many users.

jeschkies commented on conf/jenkins/configuration.yaml:
```diff
@@ -0,0 +1,55 @@
+            agentCommandStyle: Windows
+            containerInfo:
+              dockerForcePullImage: false
+              dockerImage: "${JENKINS_WINDOWS_AGENT_IMAGE:-mesosphere/jenkins-windows-node:latest}"
```
I think we should check in the [Dockerfile](https://github.com/mesosphere/jenkins-mesos-plugin/blob/master/nodes/Dockerfile.windows) for the Windows image.

kaiwalyajoshi replied:
We kept the images seperate from this repo so far: https://github.com/mesosphere/dcos-jenkins-dind-agent

Should we do the same for the Windows image if it gets complicated down the line?

jeschkies replied:
What is the benefit of this? Any new engineer will have to dig through our repo to find it.

jeschkies commented on scripts/run.sh:
```diff
@@ -1,13 +1,28 @@
+    echo "No additional plugins requested for installation..."
+fi
+
 export LD_LIBRARY_PATH=/libmesos-bundle/lib:/libmesos-bundle/lib/mesos:$LD_LIBRARY_PATH
```
This should not be required anymore. The new plugin based on USI does not use `libmesos`.

jeschkies commented on scripts/run.sh:
```diff
@@ -1,13 +1,28 @@
+
 export LD_LIBRARY_PATH=/libmesos-bundle/lib:/libmesos-bundle/lib/mesos:$LD_LIBRARY_PATH
-export JENKINS_SLAVE_AGENT_PORT=$PORT_AGENT
 export MESOS_NATIVE_JAVA_LIBRARY=$(ls /libmesos-bundle/lib/libmesos-*.so)
```
This can go.

jeschkies commented on universe/marathon.json.mustache:
```diff
@@ -6,52 +6,77 @@
       "MARATHON_NAME": "{{service.marathon-name}}",
-      {{#security.secret-name}}
+      {{#service.security.secret-name}}
         "DCOS_SERVICE_ACCOUNT_CREDENTIAL": { "secret": "serviceCredential" },
```
This should be `DCOS_SERVICE_ACCOUNT_PRIVATE_KEY`. And the `DCOS_SERVICE_ACCOUNT` option is missing as well. I'm not sure we still need `MESOS_AUTHENTICATEE` and `MESOS_HTTP_AUTHENTICATEE`.

jeschkies reviewed (changes requested):
This is pretty awesome. I have a few comments on the configuration.

kaiwalyajoshi committed:
Remove dependency on libmesos.

kaiwalyajoshi committed:
Strict Mode Packaging for USI based Jenkins. (#316)

* WIP: This works, but is messy. Make it easier and more elegant.

* Remove DCOS_SERVICE_ACCOUNT_CREDENTIAL and associated references.

* Re-add service-account, should'nt have been removed before.

* Make private-key free-form as user config can be difficult to predict.

* Use https for mesos leader URL.

* Remove debugging outputs.

kaiwalyajoshi committed:
Move jenkins cpu and memory requirements into the jenkins-master config section.

jeschkies reviewed (approved):
Nice 🎉 

Could we import the Dockerfiles for the Dind images and deprecated the other repo?

kaiwalyajoshi committed:
Add Windows and Linux agent Dockerfiles. (#317)

kaiwalyajoshi commented:
I've added both the linux and windows Dockerfiles to build the agent images.
I'll cut a beta release from this branch and soak that, we'll merge any fixes we find from soaking.

jeschkies committed:
Define default entrypoints for Docker images. (#319)

If the Docker in Docker images define the `wrapper.sh` as their default
entrypoint we won't have to use the custom shell in Mesos.

JIRA issues: DCOS_OSS-5937

kaiwalyajoshi committed:
Use /login as the health-check endpoint to allow least priveledged users.

kaiwalyajoshi committed:
MWT-24 Fixes (#320)

* Update marathon.json.mustache to use strict-mode to set security related settings.

* update jenkins.py to reflect changed options with version 4.0.0

* Change generator-job to be only run on labels of type linux as its the first job run.

* Add support for Marathon group manipulation in sdk_marathon.py

* Update sdk_security to provision secrets for new USI model of extracting secrets for version 4.0.0

* Remove provisioning of Jenkins Slaves, this is done by default when the service is installed. Update secret name to conform with USI. Add support for mesos-agent-labels.

* Mark linux node as a priviledge Dind agent. (#321)

* Mark linux node as a priviledge Dind agent.

* Change Dind image.

* Change DinD image to mesosphere/jenkins-dind:0.8.0

Co-authored-by: Kaiwalya Joshi <kjoshi@mesosphere.com>

* Add anti-affinity for os:windows Mesos Attribute by default. (#322)

* Set Jenkins linux agent to use mesosphere/jenkins-dind:0.8.0

* Change default agent user to nobody.

* Change default agent user to nobody.

* Define entrypoiht for Windows container.

* Set agent user in load test. (#323)

* Fix error when adding agent_user to options. Make service_user and agent_user configureable.

* Make service-accounts configureable for service_user and agent_user (#324)

* Set permissions for `nobody`
* Make service accounts configureable.

Co-authored-by: Kaiwalya Joshi <kjoshi@mesosphere.com>

* Set run-delay to 10 to match work-duration.

Co-authored-by: Karsten Jeschkies <kjeschkies@mesosphere.com>
Co-authored-by: Karsten Jeschkies <k@jeschkies.xyz>

kaiwalyajoshi committed:
Use version tagged releases of mesosphere/jenkins-windows-node from now on.

jeschkies committed:
Set cmd as entrypoint for Windows node container. (#325)

Summary:
The command defined in the Jenkins plugin is not compatible with
Powershell. We need to use `cmd /k` instead.

kaiwalyajoshi committed:
Bump min DC/OS version to 1.13.

kaiwalyajoshi committed:
Bump min DC/OS version to 1.11.

kaiwalyajoshi committed:
[D2IQ-67992]  Remove Windows Support. (#326)

* Remove Windows Agents from JCasC.

* Remove jenkins-agent-images.

* Remove Windows Agent Options from Marathon config and Cosmos config.

* Remove os-anti-affinity option required for Windows agents interoperability.

kaiwalyajoshi committed:
Use mesos:2.0-beta19 plugin.

kaiwalyajoshi committed:
Fixes from the Scale Test Dry-Run (#327)

- Delete any pre-existing secrets before uploading new ones.
- Bump DinD image to `19.03-dind`

kaiwalyajoshi committed:
Use mesos:2.0-beta20 plugin.

kaiwalyajoshi committed:
Use overlay2 as the storage-driver for DinD image. (#328)

kaiwalyajoshi committed:
Use Mesos 2.0 plugin.

kaiwalyajoshi committed:
Enforce Jenkins Agent Port Settings.

kaiwalyajoshi committed:
Merge branch '4.x.y' of github.com:mesosphere/dcos-jenkins-service into 4.x.y

harryge00 commented:
Exited to see this PR 🚀  when will Jenkins 2.204.2 LTS be available in Universe?

kaiwalyajoshi committed:
Fix Groovy initilization errors.

kaiwalyajoshi committed:
Fix startup failures. Bump Jenkins base-tech to 2.204.6 and bump associated plugins.

kaiwalyajoshi committed:
Set dind image to 0.9.0

kaiwalyajoshi committed:
Merge branch 'master' into 4.x.y
